packages/blockchain-link/src/workers/solana/index.ts

```diff
@@ -400,78 +471,96 @@ const subscribeAccounts = async (
                 subscribedAccount => account.descriptor === subscribedAccount.descriptor,
             ),
     );
-    newAccounts.forEach(a => {
-        const subscriptionId = api.onAccountChange(new PublicKey(a.descriptor), async () => {
-            // get the last transaction signature for the account, since that wha triggered this callback
-            const lastSignature = (
-                await api.getSignaturesForAddress(new PublicKey(a.descriptor), {
-                    before: undefined,
-                    limit: 1,
-                })
-            )[0]?.signature;
-            if (!lastSignature) return;
-
-            // get the last transaction
-            const lastTx = (
-                await api.getParsedTransactions([lastSignature], {
-                    maxSupportedTransactionVersion: 0,
-                    commitment: 'finalized',
-                })
-            )[0];
-
-            if (!lastTx || !isValidTransaction(lastTx)) {
-                return;
-            }
-
-            const tokenMetadata = await getTokenMetadata();
-            const tx = solanaUtils.transformTransaction(lastTx, a.descriptor, [], tokenMetadata);
-
-            // For token accounts we need to emit an event with the owner account's descriptor
-            // since we don't store token accounts in the user's accounts.
-            const descriptor =
-                findTokenAccountOwner(state.getAccounts(), a.descriptor)?.descriptor ||
-                a.descriptor;
-
-            post({
-                id: -1,
-                type: RESPONSES.NOTIFICATION,
-                payload: {
-                    type: 'notification',
-                    payload: {
-                        descriptor,
-                        tx,
-                    },
-                },
-            });
-        });
-        state.addAccounts([{ ...a, subscriptionId }]);
-    });
+    await Promise.all(
+        newAccounts.map(async a => {
+            const abortController = new AbortController();
+            const accountNotifications = await api.rpcSubscriptions
+                .accountNotifications(address(a.descriptor))
+                .subscribe({ abortSignal: abortController.signal });
+            const subscriptionId = NEXT_ACCOUNT_SUBSCRIPTION_ID++;
+            ACCOUNT_SUBSCRIPTION_ABORT_CONTROLLERS.set(subscriptionId, abortController);
+            const account: SubscriptionAccountInfo = {
+                ...a,
+                subscriptionId,
+            };
+            state.addAccounts([account]);
+            (async () => {
+                // TODO: Wrap this for/await loop in a try/catch and write code to recover in the event
+                // that the account subscription going down.
+                for await (const _ of accountNotifications) {
```

Comment:
The subscriptions API vends `AsyncIterables` now. You can iterate over them, wrap the whole thing in `try/catch` and do optional cleanup in `finally`. When you call `abortController.abort()` the loop will gracefully exit and enter the `finally` block.